core/commands/pin/remotepin.go

```diff
@@ -0,0 +1,714 @@
+	"github.com/libp2p/go-libp2p-core/host"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	"github.com/multiformats/go-multiaddr"
+)
+
+var log = logging.Logger("core/commands/cmdenv")
+
+var remotePinCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Pin (and unpin) objects to remote pinning service.",
+	},
+
+	Subcommands: map[string]*cmds.Command{
+		"add":     addRemotePinCmd,
+		"ls":      listRemotePinCmd,
+		"rm":      rmRemotePinCmd,
+		"service": remotePinServiceCmd,
+	},
+}
+
+var remotePinServiceCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Configure remote pinning services.",
+	},
+
+	Subcommands: map[string]*cmds.Command{
+		"add": addRemotePinServiceCmd,
+		"ls":  lsRemotePinServiceCmd,
+		"rm":  rmRemotePinServiceCmd,
+	},
+}
+
+const pinNameOptionName = "name"
+const pinCIDsOptionName = "cid"
+const pinStatusOptionName = "status"
+const pinServiceNameOptionName = "service"
+const pinServiceURLOptionName = "url"
+const pinServiceKeyOptionName = "key"
+const pinServicePinCountOptionName = "pin-count"
+const pinBackgroundOptionName = "background"
+const pinForceOptionName = "force"
+
+type RemotePinOutput struct {
+	RequestID string
+	Status    string
+	Cid       string
+	Name      string
+	Delegates []string // multiaddr
+}
+
+func toRemotePinOutput(ps pinclient.PinStatusGetter) RemotePinOutput {
+	return RemotePinOutput{
+		RequestID: ps.GetRequestId(),
+		Name:      ps.GetPin().GetName(),
+		Delegates: multiaddrsToStrings(ps.GetDelegates()),
+		Status:    ps.GetStatus().String(),
+		Cid:       ps.GetPin().GetCid().String(),
+	}
+}
+
+func printRemotePinDetails(w io.Writer, out *RemotePinOutput) {
+	tw := tabwriter.NewWriter(w, 0, 0, 1, ' ', 0)
+	defer tw.Flush()
+	fw := func(k string, v string) {
+		fmt.Fprintf(tw, "%s:\t%s\n", k, v)
+	}
+	fw("RequestID", out.RequestID)
+	fw("CID", out.Cid)
+	fw("Name", out.Name)
+	fw("Status", out.Status)
+}
+
+// remote pin commands
+
+var pinServiceNameOption = cmds.StringOption(pinServiceNameOptionName, "Name of the remote pinning service to use.")
+
+var addRemotePinCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline:          "Pin objects to remote storage.",
+		ShortDescription: "Stores an IPFS object(s) from a given path to a remote pinning service.",
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("ipfs-path", true, false, "Path to object(s) to be pinned."),
+	},
+	Options: []cmds.Option{
+		cmds.StringOption(pinNameOptionName, "An optional name for the pin."),
+		pinServiceNameOption,
+		cmds.BoolOption(pinBackgroundOptionName, "Add to the queue on the remote service and return RequestID immediately.").WithDefault(false),
+	},
+	Type: RemotePinOutput{},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		ctx, cancel := context.WithCancel(req.Context)
+		defer cancel()
+
+		// Get remote service
+		var c *pinclient.Client
+		if service, serviceFound := req.Options[pinServiceNameOptionName]; serviceFound {
```

Comment:
For all of these functions we need to bail if the pinning service name isn't passed in because there's nothing meaningful for us to do.

Options:
1. Make this an argument instead of a flag
    - It's accurate, but not copy-paste portability when we try and unify the remote and local pinning apis
    - Not great if we wanted to allow using multiple services in the same command
2. Make this a flag that we fail if we don't pass
    - Not really an "option", which may be a bit confusing
3. Have a default remote pinning service that we use
    - Would require modifications to `ipfs config` so that we only disallow showing + editing `Pinning.RemoteServices.API` and we could see `Pinning.RemoteServices.Default`
    - Might be convenient for users, but could also make transitioning to the unified API more difficult since users who may never have used used the `--service` flag.

2 seems like the easiest to implement right now even if a bit confusing the first time people run the command and we can always allow for 3 later.

@petar @lidel WDYT? I'll probably push out a commit implementing 2 soon

Reply:
Implemented 2 in 7d63db36979fab2fb43736945834330da315b1cd.

We could probably make the pinning service HTTP API client not even allow you to construct the struct and instead return an interface, but having the equivalent of null reference exception errors from not using constructors seems relatively fair game in Go as it means there was a programmer error instead of user error.

Reply:
I think (2) resolves the problem at hand, thanks!